proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,531 @@
+storage because this is how admission is enforced (which would be required if the admin wanted to give a user the
+ability to create client secrets but only for specific clients - this limitation is because the name of a new object
+may not be known at authorization time).
+
+The OIDC client secret request API will support open API schema docs to make sure commands such as `kubectl explain
+oidcclientsecretrequests.spec` work (this is not true for token credential request today).
+
+##### Disallowing audience confusion
+
+Who decides the names of the dynamic client and the workload clusters?
+
+- The name of the dynamic client would be chosen by the admin of the Supervisor.
+- The name of the workload cluster is chosen by the admin of the workload cluster (potentially a different person or
+  automated process). We don’t currently limit the string which chooses the audience name for the workload cluster, so
+  it can be any non-empty string. The Supervisor is not aware of these strings in advance.
+
+Given that, there are two kinds of potential audience confusion for the token exchange.
+
+1. The ID token issued during the original authorization flow (before token exchange) will have an audience set to the
+   name of the dynamic client. If this client’s name happened to be the same name as the name of a workload cluster,
+   then the client could potentially skip the token exchange and use the original ID token to access the workload
+   cluster via the Concierge (acting as the user who logged in). These initial ID tokens were not meant to grant access
+   to any particular cluster.  An admin may even deliberately try use dynamic clients to represent clusters as is
+   traditionally done with Dex.
+
+2. A user can use the public `pinniped-cli` client to log in and then to perform a token exchange to any audience value.
+   They could even ask for an audience which matches the name of an existing dynamic client to get back an ID token that
+   would appear as if it were issued to that dynamic client. Then they could try to find a way to use that ID token with
+   a webapp which uses that dynamic client to authenticate its users (although that may not be possible depending on the
+   implementation of the webapp).
+
+To address all of these issues we will:
+
+- Require a static, reserved prefix (`client.oauth.pinniped.dev-`) for all dynamic OAuth client IDs
+- Require a static, reserved prefix (`token-exchange.oauth.pinniped.dev-`) for audiences requested via the token
+  exchange API
+- All ID tokens issued by the supervisor will contain the `azp` claim and its value will always be set to the client ID
+  of the client that was used for the initial login flow.  This is meant to prevent any information loss during flows
+  such as the token exchange API.
+- In the ID token retrieved from the authorization flow (and only this ID token), nest the `username` and `groups`
```

Comment:
> I think it will make our code [more confusing]

The code change will be trivial so I do not consider this as a valid argument against.

> and our tokens more confusing.

They will certainly look weird but passport JS and oauth2-proxy can easily consume nested claims.  The alternative is to shove all the claims behind a user info endpoint, which is more work to implement for us and maybe more work for a web app to consume.

> My preference would be to be as clear as we can in docs but not impose any hard limitation.

Just as we are deprecating and removing the http port to impose a hard limitation on your network setup, I believe we should do the same with how you configure and consume our auth APIs.  Insecure configs should simply not be possible.

Reply:
Thought: what breaks if only the initial authorization code flow returns an ID token (i.e. it is not returned on refresh flows)?  We would be OIDC spec compliant but incompatible with the `kubectl` + API server + Dex + Gangway style approach.  I _think_ the only thing that would be missing for a web app is group membership updates?